### PR TITLE
Add Rule - Only Signed in user can create items on his wishlist

### LINKED ISSUE #15: Add Rule - Only Signed in user can create items on his wishlist
Set a rule that only logged in user can create items on his wishlist.

### PR DESCRIPTION
**Pull request by nodzy**
Closes #15

### CODE REVIEW

File: app/controllers/items_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class ItemsController < ApplicationController
+  load_and_authorize_resource :wishlist
+  load_and_authorize_resource :item, :through => :wishlist
+  before_action :authenticate_user!
+  before_action :set_wishlist!
+
+  def new
+    @item = @wishlist.items.new
+  rescue CanCan::AccessDenied
+    redirect_to wishlists_path, alert: t("generic_error")
```
Comment (student-tomasz) [inline app/controllers/items_controller.rb]: I'd redirect to the wishlist page the gift was being added to. If the wishlist is unavailable as well, then the user will be redirected to the index page.

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    @item.save!
+    redirect_to wishlists_path, notice: t(".success")
+  rescue CanCan::AccessDenied
+    redirect_to wishlists_path, alert: t("generic_error")
+  rescue ActiveRecord::RecordInvalid
+    render :new
+  end
+
+  def show
+  end
```
Comment (student-tomasz) [inline app/controllers/items_controller.rb:40]: Why did you add an empty action?

File: app/controllers/wishlists_controller.rb [hunk 3]
```diff
@@ -10,16 +10,19 @@ def index
...
-      redirect_to wishlists_path, notice: t(".success")
-    else
-      render :new
-    end
+    @wishlist.save!
+    redirect_to wishlists_path, notice: t(".success")
+  rescue CanCan::AccessDenied
+    redirect_to wishlists_path, alert: t("generic_error")
+  rescue ActiveRecord::RecordInvalid
+    render :new
```
Comment (student-tomasz) [inline app/controllers/wishlists_controller.rb]: 👍

File: config/locales/en.yml [hunk 4]
```diff
@@ -8,3 +8,6 @@ en:
       success: "Wishlist was successfully deleted."
     update:
       success: "Wishlist was successfully updated."
+  items:
+    create:
+      success: "Item was successfully created!"
```
Comment (agalkiewicz) [inline config/locales/en.yml]: Shame on You. :D Set Your editor properly to make an empty line in the end of file.

File: app/views/wishlists/index.html.erb [hunk 5]
```diff
@@ -3,7 +3,7 @@
   <ul>
     <% @wishlists.each do |wishlist| %>
       <li>
-        <%= link_to wishlist.name, "#" %>
+        <%= link_to wishlist.name, wishlist_path(wishlist) %>
```
Comment (student-tomasz) [inline app/views/wishlists/index.html.erb]: Just `wishlist` will suffice.

File: app/controllers/items_controller.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+class ItemsController < ApplicationController
+  load_and_authorize_resource :wishlist
+  load_and_authorize_resource :item, :through => :wishlist
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb]: Hmm, I will use load too.

File: config/locales/en.yml (hunk 4 above)
Comment (student-tomasz) [inline config/locales/en.yml]: Add an empty line at the end of the file.

File: spec/controllers/items_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      end
+
+      context "with invalid parameters" do
+        before { post_create(name: nil) }
+
+        include_examples "redirects to the 'sign_in' page"
+      end
+    end
+  end
+end
```
Comment (student-tomasz) [inline spec/controllers/items_controller_spec.rb]: Indentation is broken. Or `private` is outside the `describe` block?

File: spec/controllers/items_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  private
+
+  def post_create(custom = {})
+    post :create, params: { item: attributes_for(:item, custom), wishlist_id: wishlist.id  }
+  end
+
+  def get_show(id)
+    get :show, params: { id: id, wishlist_id: wishlist.id }
+  end
```
Comment (student-tomasz) [inline spec/controllers/items_controller_spec.rb:247]: Copy pasta?

File: app/controllers/items_controller.rb (hunk 2 above)
Comment (student-tomasz) [inline app/controllers/items_controller.rb:40]: I mean, we don't plan on showing single items on their own.

File: app/controllers/items_controller.rb (hunk 6 above)
Comment (student-tomasz) [inline app/controllers/items_controller.rb]: Yeah, well. The item gets loaded, and authorized automagically, and yet, you still create your own items in `#new` and `#create` and don't authorize the ones you've created manually. Use the `authorize!` manually.

File: config/routes.rb [hunk 9]
```diff
@@ -1,6 +1,8 @@
 Rails.application.routes.draw do
   root to: "wishlists#index"
-  resources :wishlists
+  resources :wishlists do
+    resources :items
```
Comment (student-tomasz) [inline config/routes.rb]: You've only added `#new` and `#create` actions in the controller. And `#show` which is not needed. Restrict the routes to those actions only, as well.

File: app/models/ability.rb [hunk 10]
```diff
@@ -4,6 +4,7 @@ class Ability
   def initialize(user)
     if user
       can %i(create read destroy update), Wishlist, user_id: user.id
+      can :manage, Item, :wishlist => { :user_id => user.id }
```
Comment (student-tomasz) [inline app/models/ability.rb]: Don't use the old hash syntax. It's literally in the line above.

Comment (agalkiewicz) [inline app/models/ability.rb]: Why :manage? You are only doing  new and create right now. But it's not so important.

Comment (nodzy) [inline app/models/ability.rb]: Guys, it has label in progress beacuse I'm still in process of refactoring :) rubocop points out array syntax and all that jazz, I'll correct everything before changing label to ready to check. Now I'm just trying to make tests to pass

File: app/controllers/items_controller.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  private
+
+  def item_params
+    params[:item] ||= {}
+    params.require(:item).permit(:wishlist_id, :name, :description)
+  end
+
+  def set_wishlist!
```
Comment (Kacper3331) [inline app/controllers/items_controller.rb]: Is this exclamation "!" necessary?

File: config/locales/en.yml (hunk 4 above)
Comment (Kacper3331) [inline config/locales/en.yml]: You forgot about empty line.

File: app/controllers/items_controller.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    @wishlist = Wishlist.find(params[:wishlist_id])
+  rescue ActiveRecord::RecordNotFound
+    redirect_to wishlists_path, alert: t("generic_error")
+  end
```
Comment (shadre) [inline app/controllers/items_controller.rb:58]: Don't we need to rescue from `CanCan::AccessDenied` here as well?

File: config/locales/en.yml (hunk 4 above)
Comment (shadre) [inline config/locales/en.yml]: What is going on here, should I point this out too? ;-)

File: spec/controllers/items_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        it "assigns an instance variable under @item" do
+          expect(assigns(:item)).to be_a(Item)
+        end
+
+        it "persists the new item" do
+          expect(assigns(:item)).to be_persisted
+          expect { post_create(wishlist_id: wishlist.id) }.to change(Item, :count).by(1)
+        end
+
+        it "returns a redirect" do
```
Comment (shadre) [inline spec/controllers/items_controller_spec.rb]: Not sure this test is needed if you then test specific redirect in the next one.

File: config/routes.rb [hunk 14]
```diff
@@ -1,7 +1,7 @@
 Rails.application.routes.draw do
   root to: "wishlists#index"
   resources :wishlists do
-    resources :items
+    resources :items, only: [:new, :create]
```
Comment (student-tomasz) [inline config/routes.rb]: `%i(new create)`

File: app/controllers/items_controller.rb (hunk 11 above)
Comment (nodzy) [inline app/controllers/items_controller.rb]: I wrote it the same way as in WishlistController which is already on master - I think it's good practise, because we would know if method throws an error, it wouldn't just fail silently :)

File: app/controllers/items_controller.rb (hunk 12 above)
Comment (nodzy) [inline app/controllers/items_controller.rb:58]: In WishlistController we rescued only RecordNotFound error in this method - I think it's not necessary, as after finding the specific wishlist, we continue to new action for item and then check for access. But I can be wrong :)

File: spec/controllers/items_controller_spec.rb (hunk 13 above)
Comment (nodzy) [inline spec/controllers/items_controller_spec.rb]: I wrote it in analogy to Wishlist controller tests (already on master), so I'd prefer to leave it to be consistent with Rspec output

File: config/routes.rb (hunk 14 above)
Comment (nodzy) [inline config/routes.rb]: Thanks, I'll correct that :+1:

File: app/controllers/items_controller.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  private
+
+  def item_params
+    params[:item] ||= {}
+    params.require(:item).permit(:wishlist_id, :name, :description)
+  end
+
+  def set_wishlist!
+    @wishlist = Wishlist.find(params[:wishlist_id])
+  rescue ActiveRecord::RecordNotFound
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb:50]: "It is okay for now, but generally it should be handled somewhere on the higher level to catch all those exceptions and render error page." - this is what Smejfu said in my pr.

File: app/controllers/items_controller.rb (hunk 11 above)
Comment (student-tomasz) [inline app/controllers/items_controller.rb]: But we catch the error inside. I don't really know, why we were told to add the `!` here.

File: app/controllers/items_controller.rb (hunk 12 above)
Comment (student-tomasz) [inline app/controllers/items_controller.rb:58]: Yeah, `CanCan::AccessDenied` won't be thrown here.

Comment (shadre) [inline app/controllers/items_controller.rb:58]: Oh right, that's true. :+1:

File: app/controllers/items_controller.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    @item = @wishlist.items.new
+    authorize! :new, @item
+  rescue CanCan::AccessDenied
+    redirect_to @wishlist, alert: t("generic_error")
+  end
+
+  def create
+    @item = @wishlist.items.create(item_params)
+    authorize! :create, @item
+    @item.save!
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb:16]: `create = new + save`
So writing `new` in the first line and `save!` here would be a valid option.

File: spec/controllers/items_controller_spec.rb (hunk 13 above)
Comment (agalkiewicz) [inline spec/controllers/items_controller_spec.rb]: I forgot to delete these tests in "user logged in" context. For mi this one is not necessary. I remember that Smejfu pointed it in my Shop code.

File: app/controllers/items_controller.rb (hunk 16 above)
Comment (nodzy) [inline app/controllers/items_controller.rb:16]: Good point, I didn't notice. Thank you

File: app/controllers/items_controller.rb (hunk 11 above)
Comment (agalkiewicz) [inline app/controllers/items_controller.rb]: I think Tomek is right. I know that I did the same in wishlists_controller_spec but I forgot to correct it.

File: spec/controllers/items_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,101 @@
+require "rails_helper"
+
+RSpec.describe ItemsController do
+  context "when user is logged in" do
+    before { sign_in_user }
+    let(:wishlist) { create(:wishlist, user: controller.current_user) }
+    let(:item) { create(:item, wishlist: wishlist) }
```
Comment (agalkiewicz) [inline spec/controllers/items_controller_spec.rb]: Is it necessary? You don't use it imo.

File: app/controllers/items_controller.rb [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    redirect_to @wishlist, alert: t("generic_error")
+  rescue ActiveRecord::RecordInvalid
+    render :new
+  end
+
+  private
+
+  def item_params
+    params[:item] ||= {}
+    params.require(:item).permit(:wishlist_id, :name, :description)
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb]: Is `:wishlist_id` necessary too?

File: app/controllers/items_controller.rb [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+  rescue CanCan::AccessDenied
+    redirect_to @wishlist, alert: t("generic_error")
+  rescue ActiveRecord::RecordInvalid
+    render :new
+  end
+
+  private
+
+  def item_params
+    params[:item] ||= {}
```
Comment (smt116) [inline app/controllers/items_controller.rb]: is that line needed?

File: app/views/wishlists/show.html.erb [hunk 20]
```diff
@@ -1,7 +1,7 @@
 <h1><%= @wishlist.name %></h1>
 
 <p><%= @wishlist.description %></p>
-
+ <%= link_to "Create new item", new_wishlist_item_path(@wishlist) %>
```
Comment (smt116) [inline app/views/wishlists/show.html.erb]: indentation

File: spec/controllers/items_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+        end
+
+        it "renders the 'new' template" do
+          expect(response).to render_template(:new)
+        end
+      end
+    end
+  end
+
+  context "when not logged in" do
```
Comment (smt116) [inline spec/controllers/items_controller_spec.rb]: for me this context should be inside method context, so:

```
describe #new
  context when user is logged in
    assertions
  context when user is logged out
    assertions
```

File: app/controllers/items_controller.rb (hunk 18 above)
Comment (nodzy) [inline app/controllers/items_controller.rb]: In WishlistController we have user_id as permitted param, I wrote it in analogy - is it wrong?

Comment (smt116) [inline app/controllers/items_controller.rb]: so you permit user to update `wishlist_id`?

Comment (student-tomasz) [inline app/controllers/items_controller.rb]: @nodzy In the `WishlistsController` I specifically added a line to overwrite the `user_id` field with the correct one. So the sent-by-user one would be dismissed and overwritten with his actual `user_id` before any actions are run. You don't do such thing here.

File: spec/controllers/items_controller_spec.rb (hunk 21 above)
Comment (student-tomasz) [inline spec/controllers/items_controller_spec.rb]: I think specs are up for an overall refactor. We'll follow through with your comment, but would like to push this, as it is, right now.

File: app/controllers/wishlists_controller.rb (hunk 3 above)
Comment (shadre) [inline app/controllers/wishlists_controller.rb:26]: Add a message for this as well. We discussed this in Ala's PR, you can see it for reference.
(And then you need to test this, too)

Comment (nodzy) [inline app/controllers/wishlists_controller.rb:26]: OK, done

**Comment by nodzy** [comment]
@Kacper3331 @agalkiewicz @shadrk @student-tomasz please take a look - I rebased and resolved conflicts, is there anything else I should change?

File: app/controllers/items_controller.rb [hunk 22]
```diff
@@ -1,7 +1,27 @@
 class ItemsController < ApplicationController
   before_action :authenticate_user!
+  before_action :set_wishlist, only: %i(new create)
   before_action :set_wishlist_and_item, only: %i(edit update)
```
Comment (student-tomasz) [inline app/controllers/items_controller.rb]: If you're extracting `set_wishlist` as another method, then change `set_wishlist_and_item` to just `set_item` and run both `set_wishlist` and `set_item` for the methods that need them both. With how things are right now you're duplicating the `set_wishlist` code unnecessarily.

Comment (nodzy) [inline app/controllers/items_controller.rb]: Ok, done, can I change label to ready_to_check?

**Comment by student-tomasz** [comment]
👍

**Comment by shadrk** [comment]
:+1:

File: spec/controllers/items_controller_spec.rb [hunk 23]
```diff
@@ -10,6 +10,106 @@
     end
   end
 
+  describe "GET #new" do
+    context "when user is logged in" do
+      before { sign_in_user }
+
+      context "when wishlist exists" do
+        before { get :new, wishlist_id: wishlist.id }
```
Comment (agalkiewicz) [inline spec/controllers/items_controller_spec.rb]: Add an empty line.

File: spec/controllers/items_controller_spec.rb [hunk 24]
```diff
@@ -10,6 +10,106 @@
...
+        it "assigns an instance variable under @item" do
+          expect(assigns(:item)).to be_a(Item)
+        end
+
+        it "persists the new item" do
+          expect(assigns(:item)).to be_persisted
+          expect { post_create(name: "name", wishlist_id: wishlist.id) }.to change(Item, :count).by(1)
+        end
+
+        it "redirects to the wishlists index page" do
```
Comment (agalkiewicz) [inline spec/controllers/items_controller_spec.rb]: Not `wishlists` - `wishlist`.

File: spec/controllers/items_controller_spec.rb (hunk 23 above)
Comment (student-tomasz) [inline spec/controllers/items_controller_spec.rb]: Should be:
```ruby
before { get :new, params: { wishlist_id: wishlist.id } }
```

Comment (student-tomasz) [inline spec/controllers/items_controller_spec.rb]: Either way, put it in a helper method.

File: spec/controllers/items_controller_spec.rb (hunk 24 above)
Comment (shadre) [inline spec/controllers/items_controller_spec.rb]: ^ Why?

File: app/controllers/items_controller.rb [hunk 25]
```diff
@@ -1,6 +1,26 @@
...
   before_action :authenticate_user!
-  before_action :set_wishlist_and_item, only: %i(edit update)
+  before_action :set_wishlist, only: %i(new create edit update)
+  before_action :set_item, only: %i(edit update)
+
+  def new
+    @item = @wishlist.items.new
+    authorize! :new, @item
+  rescue CanCan::AccessDenied
+    redirect_to root_url, alert: t("generic_error")
```
Comment (student-tomasz) [inline app/controllers/items_controller.rb]: I'd redirect to the wishlists being edited, not root. Here and in other `redirect`s in the `ItemsController`.

Comment (nodzy) [inline app/controllers/items_controller.rb]: should I change edit/update which weren't part of this issue?

Comment (student-tomasz) [inline app/controllers/items_controller.rb]: Eh, I guess so. When it makes sense.

Comment (nodzy) [inline app/controllers/items_controller.rb]: I wonder where to redirect in set_wishlist and set_item - when wishlist or item isn't found, I suppose we should redirect to root?

Comment (student-tomasz) [inline app/controllers/items_controller.rb]: `set_wishlist` -> `wishlists#index`, `set_item` -> `wishlists#show`

Comment (nodzy) [inline app/controllers/items_controller.rb]: ok, done

**Comment by student-tomasz** [comment]
You could extract the duplicated redirects tests in specs into a `shared_examples` but 👍

**Comment by nodzy** [comment]
It was a good point, I refactored it.

File: app/controllers/items_controller.rb [hunk 26]
```diff
@@ -1,19 +1,39 @@
...
   before_action :authenticate_user!
-  before_action :set_wishlist_and_item, only: %i(edit update)
+  before_action :set_wishlist, only: %i(new create edit update)
+  before_action :set_item, only: %i(edit update)
+
+  def new
+    @item = @wishlist.items.new
+    authorize! :new, @item
+  rescue CanCan::AccessDenied
+    redirect_to @wishlist, alert: t("generic_error")
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: IMO it's not necessary to handle this exception in other way than default one. You try to create a new item on not your own wishlist - redirection to root_url. You're not logged in - redirection to root_url.

File: app/controllers/items_controller.rb [hunk 27]
```diff
@@ -1,19 +1,39 @@
...
+  end
+
+  def create
+    @item = @wishlist.items.new(item_params)
+    authorize! :create, @item
+    @item.save!
+    redirect_to @wishlist, notice: t(".success")
+  rescue CanCan::AccessDenied
+    redirect_to @wishlist, alert: t("generic_error")
```
Comment (agalkiewicz) [inline app/controllers/items_controller.rb:19]: The same as above.

File: app/controllers/items_controller.rb (hunk 26 above)
Comment (student-tomasz) [inline app/controllers/items_controller.rb:10]: Yeah, we should catch this error in one place only. Don't really know how to achieve this. Separate issue? I'd let it stay here the way it is for now.

Comment (nodzy) [inline app/controllers/items_controller.rb:10]: Ok, I'll leave it for now

Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: I don't understand. It's only about deleting this error handling.

Comment (student-tomasz) [inline app/controllers/items_controller.rb:10]: I've thought it through. Yeah, you're right. Both rescues ultimately lead to the same thing. But shouldn't we handle them differently? I mean, we're not logging those errors, so for us this doesn't exactly matter right now, why user gets redirected. I dunno, feels like a good precaution though, to catch it here explicitly. Alas I can't think of a solid reason for doing so. I'd leave it, and refactor the whole catching `CanCan::AccessDenied` everywhere in another issue tomorrow?

Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: Ok

Comment (gotar) [inline app/controllers/items_controller.rb:10]: in Application controller you can:

``` ruby
class ApplicationController < ActionController::Base
  rescue_from ActionController::RoutingError, :with => :error_render_method

  def error_render_method
    respond_to do |type|
      type.xml { render :template => "errors/error_404", :status => 404 }
      type.all  { render :nothing => true, :status => 404 }
    end
    true
  end
end
```

and catch all there (ps for cancan there is already global method (helper) to catch all errors in one place)

from cancan doc: 

``` ruby
class ApplicationController < ActionController::Base
  rescue_from CanCan::AccessDenied do |exception|
    redirect_to main_app.root_url, :alert => exception.message
  end
end
```

**Comment by gotar** [comment]
ok you can merge (I put some comment)

File: app/controllers/items_controller.rb (hunk 26 above)
Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: We have used this helper. This is why I put such a comment above. :)

Comment (nodzy) [inline app/controllers/items_controller.rb:10]: I'm still wondering - maybe we should distinguish these two options. When someone tries to create/edit someone else's wishlist is redirected to this specific wishlist show page (page is in fact re-rendered). Maybe it's better then being redirected to root which is default?

Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: Create someone else's wishlist? How can it be done? Try to edit someone else's wishlist and be redirected to this specific wishlist? Which one? Someone else's??

Comment (nodzy) [inline app/controllers/items_controller.rb:10]: I meant item on wishlist, sorry :)

Comment (agalkiewicz) [inline app/controllers/items_controller.rb:10]: I think that then `ActiveRecord::RecordNotFound` would by raised - when try to edit.
